### PR TITLE
Node informational states (#4)

### PR DESCRIPTION
**Pull request by OliverEGreen**
### Purpose

This PR introduces a new design to the InfoBubble UserControl, which is used to report node warnings and errors to the user. 

There are now 3 kinds of user-facing messages a node can display. In order of ascending severity, these are Info, Warning and Message. 

The InfoBubble is now capable of displaying multiple info messages, warnings and errors simultaneously, and gives the user the ability to dismiss both Warnings and Info Messages. Errors cannot be dismissed. 

Dismissed messages may be retrieved by the node's context menu.

The following GIFs demonstrate the new functionality added in this PR:

DismissingMessages

1. User-facing messages may now be dismissed by the user, and undismissed using the node's context menu. When messages are dismissed, a small counter glyph appears at the base of the node whose messages were dismissed. 

StackingMessages

2. Messages may be stacked according to their severity level. Errors (if any) will always appear at the base of the stack, with warnings above and info messages at the highest level. Each level of the stack may be expanded and contracted by clicking on the message body.

The message body may take 1 of 3 states:

- A simple icon, with no messages displayed. This is the most compact representation.
- An icon alongside the first message at the corresponding severity level, e.g. Errors. Where there are multiple messages at this level, an iterating counter appears beside each message.
- Where multiple messages are present at the same severity level, clicking 'show all' will expand all messages at this level. An interating counter will be displayed to help users keep track of the messages.

ScrollingMessages

3. A maximum height has been set per message-level, in case warning messages are very long (such as a long Python traceback message). In such cases, a scrollbar appears to allow for convenient traversal and review of any messages. 

### Impact assessment

- [ ] Changes the way 3rd party nodes work (Add ikeough to the reviewer list)
- [ ] Includes migrations (Add ikeough to the reviewer list)
- [ ] Adds or changes a public interface (Add lukechurch to the reviewer list)
- [ ] API change is added to the [API Changes document](https://github.com/DynamoDS/Dynamo/wiki/API-Changes).
- [ ] Affects which components of the system depends on which other components (Add pboyer to the reviewer list)
- [ ] Introduces new components (Add lukechurch to the reviewer list)
- [ ] Has a performance impact (Add keyu to the reviewer list)
- [ ] Changes the installer (Add sharadkjaiswal to the reviewer list)
- [X] User facing strings, if any, are extracted into `*.resx` files
- [X] All tests pass using the self-service CI.
- [X] Snapshot of UI changes, if any.

#### References (FILL ME IN)

The Dynamo team already has links to the new UI design.

#### Testing strategy

These changes have been tested manually in DynamoSandbox, with a range of generated Info Messages, Warnings and Errors.

### Review staging

- [ ] Architectural Design Review
- [X] UI Review

### Reviewers

@QilongTang 

### FYIs

@Amoursol

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs [hunk 1]
```diff
@@ -56,6 +57,7 @@ public partial class NodeViewModel : ViewModelBase
         private bool isexplictFrozen;
         private bool canToggleFrozen = true;
         private bool isRenamed = false;
+        private ObservableCollection<MenuItem> dismissedAlerts = new ObservableCollection<MenuItem>();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs:60]: hmm.. I think it is usually not recommended to referencing or dynamically manage view level controls in view model. Is there a workaround of this, using binding for example so we can keep the `ObservableCollection<MenuItem> dismissedAlerts` on node view level but bind to a collection of messages on nodeviewModel maybe?

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs [hunk 2]
```diff
@@ -50,56 +54,157 @@ public enum State
...
+        {
+            get => nodeErrorsShowLessMessageVisible;
+            set
+            {
+                nodeErrorsShowLessMessageVisible = value;
+                RaisePropertyChanged(nameof(NodeErrorsShowLessMessageVisible));
+            }
+        }
+
+        public string Content
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:148]: Let's add comments for this

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs [hunk 3]
```diff
@@ -50,56 +54,157 @@ public enum State
...
+        {
+            get { return content; }
+            set { content = value; RaisePropertyChanged("Content"); }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:151]: Let's use nameof

File: src/DynamoCore/DynamoCore.csproj [hunk 4]
```diff
@@ -95,6 +95,7 @@ limitations under the License.
     <Reference Include="System.Runtime.Serialization" />
     <Reference Include="System.Xml" />
     <Reference Include="System.Xml.Linq" />
+    <Reference Include="WindowsBase" />
```
Comment (QilongTang) [inline src/DynamoCore/DynamoCore.csproj:98]: Just sanity check this is indeed need for?

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs [hunk 5]
```diff
@@ -153,12 +416,27 @@ public InfoBubbleViewModel(DynamoViewModel dynamoViewModel)
             ConnectingDirection = Direction.None;
             Content = string.Empty;
             DocumentationLink = null;
-            ZIndex = 3;
+            // To appear above any NodeView elements
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:419]: Just curious, how to find this fact out?

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs [hunk 6]
```diff
@@ -153,12 +416,27 @@ public InfoBubbleViewModel(DynamoViewModel dynamoViewModel)
...
+            ZIndex = 7; 
             InfoBubbleStyle = Style.None;
             InfoBubbleState = State.Minimized;
+
+            NodeMessages.CollectionChanged += NodeInformation_CollectionChanged;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:424]: Seems we need to unsubscribe somewhere?

**Review by QilongTang — COMMENTED** [review]
Look solid with some comments

File: src/DynamoCore/DynamoCore.csproj (hunk 4 above)
Comment (mjkkirschner) [reply src/DynamoCore/DynamoCore.csproj:98]: probably Application class
https://docs.microsoft.com/en-us/dotnet/api/system.windows?view=net-5.0

Comment (mjkkirschner) [reply src/DynamoCore/DynamoCore.csproj:98]: it looks like this binary is actually cross platform despite it's name - but I would also like to know why we need it - we should avoid adding unnecessary dependencies, especially to DynamoCore.

**Comment by QilongTang** [comment]
Once everything is good on https://github.com/DynamoDS/Dynamo/pull/11933, we will merge that first, then this one. We can maybe leave the comments to be addressed in separate PR.

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs (hunk 2 above)
Comment (OliverEGreen) [reply src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:148]: This is a property from the old code. There's a lot of obsolete code in the ViewModel now - I'll flag it as such.

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs (hunk 3 above)
Comment (OliverEGreen) [reply src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:151]: Same as above

File: src/DynamoCore/DynamoCore.csproj (hunk 4 above)
Comment (OliverEGreen) [reply src/DynamoCore/DynamoCore.csproj:98]: Removed this and encountered no issues.

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs (hunk 5 above)
Comment (OliverEGreen) [reply src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:419]: There was a comment I read somewhere; I think groups were at 0, connectors at 1 and node bodies at 2. I may be wrong here though.

**Comment by QilongTang** [comment]
@OliverEGreen I can't confirm but I think the `Dynamo SelfServe` job timed out because of the WPFCLI test time out so we never got a result back. What I am going to do is to merge this PR as well, and then check out the regressions as well. Let's address the comments in a different PR. What do you think?

**Comment by QilongTang** [comment]
Merging for now and @OliverEGreen Should be able to address comments in a follow up PR

File: src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs [hunk 7]
```diff
@@ -695,8 +707,76 @@ public NodeViewModel(WorkspaceViewModel workspaceViewModel, NodeModel logic)
             ++NoteViewModel.StaticZIndex;
 
             if(workspaceViewModel.InCanvasSearchViewModel.TryGetNodeIcon(this, out ImageSource imgSource)) ImageSource = imgSource;
+
+            // The Node displays a count of dismissed messages, listening to that collection in the node's ErrorBubble
+            ErrorBubble.DismissedMessages.CollectionChanged += DismissedNodeWarnings_CollectionChanged;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs:712]: @QilongTang @OliverEGreen @SHKnudsen can we unsubscribe these handlers in the NodeViewModel.Dispose please?

File: src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs (hunk 6 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:424]: agree, was this ever taken care of?

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Preview/InfoBubbleViewModel.cs:424]: @mjkkirschner Thanks for checking these out! @OliverEGreen was away and should be able to address these comments on a new PR

**Pull request merged**